src/hotspot/share/opto/reachability.cpp

```diff
@@ -0,0 +1,555 @@
+    assert(rf->outcnt() > 0, "dead node");
+    if (ignored_rfs.member(rf)) {
+      continue; // skip
+    } else if (is_redundant_rf(rf, rf_only)) {
+      dump_rfs_on(tty, this, ignored_rfs, rf_only);
+      return true;
+    }
+  }
+  return false;
+}
+#endif // ASSERT
+
+
+//======================================================================
+//---------------------------- Phase 1 ---------------------------------
+// Optimization pass over reachability fences during loop opts.
+// Eliminate redundant RFs and move RFs with loop-invariant referent out of the loop.
+bool PhaseIdealLoop::optimize_reachability_fences() {
+  Compile::TracePhase tp(_t_reachability_optimize);
+
+  assert(OptimizeReachabilityFences, "required");
+
+  Unique_Node_List redundant_rfs;
+  find_redundant_rfs(redundant_rfs);
+
+  Node_List worklist;
+  for (int i = 0; i < C->reachability_fences_count(); i++) {
+    ReachabilityFenceNode* rf = C->reachability_fence(i);
+    if (!redundant_rfs.member(rf)) {
+      // Move RFs out of counted loops when possible.
+      IdealLoopTree* lpt = get_loop(rf);
+      Node* referent = rf->referent();
+      Node* loop_exit = lpt->unique_loop_exit_or_null();
+      if (lpt->is_invariant(referent) && loop_exit != nullptr) {
+        // Switch to the outermost loop.
+        for (IdealLoopTree* outer_loop = lpt->_parent;
+             outer_loop->is_invariant(referent) && outer_loop->unique_loop_exit_or_null() != nullptr;
+             outer_loop = outer_loop->_parent) {
+          assert(is_member(outer_loop, rf), "");
+          loop_exit = outer_loop->unique_loop_exit_or_null();
+        }
+        assert(loop_exit != nullptr, "");
+        worklist.push(referent);
+        worklist.push(loop_exit);
+        redundant_rfs.push(rf);
+      }
+    }
+  }
+
+  // Populate RFs outside counted loops.
+  while (worklist.size() > 0) {
+    Node* ctrl_out = worklist.pop();
+    Node* referent = worklist.pop();
+    insert_rf(ctrl_out, referent);
+  }
+
+  // Redundancy is determined by dominance relation.
+  // Sometimes it becomes evident that an RF is redundant once it is moved out of the loop.
+  // Also, newly introduced RF can make some existing RFs redundant.
+  find_redundant_rfs(redundant_rfs);
+
+  // Eliminate redundant RFs.
+  bool progress = (redundant_rfs.size() > 0);
+  while (redundant_rfs.size() > 0) {
+    remove_rf(redundant_rfs.pop()->as_ReachabilityFence());
+  }
+
+  assert(redundant_rfs.size() == 0, "");
+  assert(!has_redundant_rfs(redundant_rfs, true /*rf_only*/), "");
+
+  return progress;
+}
+
+//======================================================================
+//---------------------------- Phase 2 ---------------------------------
+
+// Linearly traverse CFG upwards starting at n until first merge point.
+// All encountered safepoints are recorded in safepoints list.
+static void linear_traversal(Node* n, Node_Stack& worklist, VectorSet& visited, Node_List& safepoints) {
```

Comment:
The second comment line does not sound accurate, we don't collect ALL, only the candidates. Maybe we can find a better method name, and even remove that comment because of it?

Given the more useful sub query `is_interfering_sfpt_candidate`, I think we could name this method something like `collect_interfering_sfpt_candidates`. Or is it very important that this is a linear traversal vs some other traversal we could choose from?

Reply:
Hmm, but this here is only a component of the `enumerate_interfering_sfpts` below, which has essencially that name.

So maybe it should be `collect_interfering_sfpt_candidates_for_node` here and just `collect_interfering_sfpt_candidates` below?

Reply:
I updated the comment and did some renaming.

`enumerate_interfering_sfpts_linear_traversal` computes the set of interfering safepoints for some RF node, not just candidates. It's only `is_interfering_sfpt_candidate()` which does some filtering ignoring the context (the RF node a safepoint can interfere with).